crates/red_knot_python_semantic/src/types/call/bind.rs

```
@@ -0,0 +1,365 @@
+            .collect(),
+        errors,
+    }
+}
+
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub(crate) struct CallBinding<'db> {
+    /// Type of the callable object (function, class...)
+    callable_ty: Option<Type<'db>>,
+
+    /// Return type of the call.
+    return_ty: Type<'db>,
+
+    /// Bound types for parameters, in parameter source order.
+    parameter_tys: Box<[Type<'db>]>,
+
+    /// Call binding errors, if any.
+    // TODO use SmallVec once variance bug is fixed
+    errors: Vec<CallBindingError<'db>>,
+}
+
+impl<'db> CallBinding<'db> {
+    // TODO remove this constructor and construct always from `bind_call`
+    pub(crate) fn from_return_ty(return_ty: Type<'db>) -> Self {
+        Self {
+            callable_ty: None,
+            return_ty,
+            parameter_tys: Box::default(),
+            errors: vec![],
+        }
+    }
+
+    pub(crate) fn set_return_ty(&mut self, return_ty: Type<'db>) {
+        self.return_ty = return_ty;
+    }
+
+    pub(crate) fn return_ty(&self) -> Type<'db> {
+        self.return_ty
+    }
+
+    pub(crate) fn parameter_tys(&self) -> &[Type<'db>] {
+        &self.parameter_tys
+    }
+
+    pub(crate) fn first_parameter(&self) -> Option<Type<'db>> {
+        self.parameter_tys().first().copied()
+    }
+
+    fn callable_name(&self, db: &'db dyn Db) -> Option<&ast::name::Name> {
+        match self.callable_ty {
+            Some(Type::FunctionLiteral(function)) => Some(function.name(db)),
+            Some(Type::ClassLiteral(class_type)) => Some(class_type.class.name(db)),
+            _ => None,
+        }
+    }
+
+    pub(super) fn report_diagnostics(&self, context: &InferContext<'db>, node: ast::AnyNodeRef) {
+        let callable_name = self.callable_name(context.db());
+        for error in &self.errors {
+            error.report_diagnostic(context, node, callable_name);
+        }
+    }
+}
+
+#[derive(Clone, Debug, PartialEq, Eq)]
+pub(crate) struct ParameterContext {
```

Can we add some documentation to this type (or better, come up with a better name). The name is fairly meaningless to me (I know it has something to do with parameter)

---

Any better name suggestions? I had trouble coming up with a good name. It's "the information we need to report an error regarding a particular parameter."

I can add a doc comment.

---

Doc comment added.